### PR TITLE
SortNotesTest

### PR DESCRIPTION
**Pull request by finarugeanina**
The sortNotesTest  will add a required number of  random notes or checklists with random titles and contents and  after it clicks on the sort by title button, it checks to see if the notes are ordered by titles.

### CODE REVIEW

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/AddSingleTextNoteTest.java [hunk 1]
```diff
@@ -57,10 +52,6 @@ public void addSingleTextNoteTest() {
         Assert.assertTrue(emptyListNotDisplayedErrorMessage, checkIfTextIsDisplayed(emptyListId, emptyListMessage));
         addNewTextNote(title, content , hour, minutes);
 
-        Assert.assertTrue(reminderSetForErrorMessage, checkIfMatches(dateTime, withSubstring(reminderSetForMessage)));
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/AddSingleTextNoteTest.java:60]: why were these removed?

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/LockedNoteTest.java [hunk 2]
```diff
@@ -60,6 +59,6 @@ public void lockedNoteTest() {
         Assert.assertTrue(lockedIconNotDisplayedMessage,checkIfMatches(lockedItemView,isDisplayed()));
 
         openNthNote(0);
-        Assert.assertTrue(NewNoteStringVariables.editPageisNotDisplayedMessage,checkIfMatches(editPageView,isDisplayed()));
+        Assert.assertTrue(editPageisNotDisplayedMessage,checkIfMatches(editPageView,isDisplayed()));
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/LockedNoteTest.java]: missing space after comma (format the code)

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/EspressoUtils.java [hunk 3]
```diff
@@ -77,6 +80,23 @@ static void clickSelectedUiObjectWithDescription(String contentDescription){
...
+            UiObject selectedDescription = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation()).findObject(new UiSelector().text(textToBeFound));
+            selectedDescription.click();
+            selectedDescription.setText(textToBeReplaced);
+        } catch (UiObjectNotFoundException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public static String getTextFromSelectedUiObject(int index) throws UiObjectNotFoundException {
+
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/EspressoUtils.java]: remove this empty line

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/NewCheckList.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import static androidx.test.espresso.matcher.ViewMatchers.withSubstring;
+import static androidx.test.espresso.matcher.ViewMatchers.withText;
+import static it.feio.android.omninotes.utils.EspressoUtils.checkIfMatches;
+import static it.feio.android.omninotes.utils.EspressoUtils.clickBack;
+import static it.feio.android.omninotes.utils.EspressoUtils.clickOnButton;
+import static it.feio.android.omninotes.utils.NewNoteStringVariables.reminderSetForErrorMessage;
+import static it.feio.android.omninotes.utils.NewNoteStringVariables.reminderSetForMessage;
+import static org.hamcrest.Matchers.allOf;
+
+class NewCheckList extends NewNote{
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/NewCheckList.java]: format the code (missing space after {)

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/NewTextNote.java [hunk 5]
```diff
@@ -25,14 +32,17 @@ private static void addContent(String selectedContent) {
     public static void addNewTextNote(String selectedTitle, String selectedContent, String selectedHour, String selectedMinutes) {
         addNewTextNote(selectedTitle,selectedContent);
         addDateTime(selectedHour, selectedMinutes);
-
         clickOnButton(okButton);
+        Assert.assertTrue(reminderSetForErrorMessage, checkIfMatches(dateTime, withSubstring(reminderSetForMessage)));
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/NewTextNote.java:49]: i would not recommend to keep the asserts inside this method. what if we want to test an error?: we add the note, hit ok and want to check the presence of an error (suppose there is somewhere a request that is made and its status code is 404). in this case, the behaviour is expected, but this method will fail because of the asserts.

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import static androidx.test.espresso.Espresso.onView;
+import static androidx.test.espresso.matcher.ViewMatchers.withId;
+import static androidx.test.espresso.matcher.ViewMatchers.withParent;
+import static androidx.test.espresso.matcher.ViewMatchers.withText;
+import static it.feio.android.omninotes.utils.EspressoUtils.*;
+import static it.feio.android.omninotes.utils.NewNoteStringVariables.titleButton;
+import static org.hamcrest.Matchers.allOf;
+
+public class SortNotes {
+    public static ArrayList<String> sortNotes(ArrayList<String> arrayWithTitles, int numberOfTimes) {
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java]: you could use List< String> here in both places

Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java]: from what i see, this method is not only sorting, it also collects the titles in a list

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/NewCheckList.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    private static ViewInteraction addNewCheckListMatcher = onView(withId(R.id.fab_checklist));
+    private static ViewInteraction okButton = onView(allOf(withId(R.id.buttonPositive), withText("OK"), withParent(withId(R.id.button_layout))));
+
+    private static void addCheckListContent(String content1, String content2) {
+        EspressoUtils.writeSelectedUiObjectWithText("New item", content1);
+        EspressoUtils.writeSelectedUiObjectWithText("New item", content2);
+    }
+
+    static void addNewCheckList(String selectedTitle, String content1, String content2, String selectedHour, String selectedMinutes) {
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/NewCheckList.java:49]: you can make all these methods public

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/SortNotesTest.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    public void deleteNotes(){
+        deleteAllNotes();
+    }
+
+    @Test
+    public void sortNotesTest() {
+        int numberOfTimes = 5;
+        ArrayList<String> arrayWithTitles = new ArrayList<>();
+
+        addMultipleRandomTextNoteOrChecklist(numberOfTimes);
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/SortNotesTest.java:44]: make sure the "creation date" criteria is selected for sorting before adding the notes

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/SortNotesTest.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+    @Before
+    public void deleteNotes(){
+        deleteAllNotes();
+    }
+
+    @Test
+    public void sortNotesTest() {
+        int numberOfTimes = 5;
+        ArrayList<String> arrayWithTitles = new ArrayList<>();
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/SortNotesTest.java]: List< String > arrayWithTitles = new ArrayList<>();

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/SortNotesTest.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+        addMultipleRandomTextNoteOrChecklist(numberOfTimes);
+
+        arrayWithTitles = new ArrayList<>(SortNotes.sortNotes(arrayWithTitles, numberOfTimes));
+        Assert.assertEquals("The list is not sorted!", arrayWithTitles.stream().sorted().collect(Collectors.toList()), arrayWithTitles);
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/SortNotesTest.java:47]: here you should work with 2 different lists: one containing the notes in the order that have been added and one containing the notes in the order resulted after sorting. when comparing, you should sort as above only the first list

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+        int index = 0;
+        ViewInteraction sortButton = onView(withId(R.id.menu_sort));
+
+        clickOnButton(sortButton);
+        clickOnButton(onView(withText(titleButton)));
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java]: when switching between sorting options, i noticed that not all the times the selection takes place. sometimes, you need to redo the action. therefore, it is necessary some kind of check here that confirms the action

**Review by rbutnariu — CHANGES_REQUESTED** [review]
try to add javadocs to your code that is non-trivial

**Comment by finarugeanina** [comment]
> try to add javadocs to your code that is non-trivial

added the javadocs

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java (hunk 11 above)
Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java]: I changed the method so it will click the button until the selection is made

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/SortNotesTest.java (hunk 10 above)
Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/SortNotesTest.java:47]: I compared the list containing the titles after the sorting by title is selected with the list sorted by a function. If they are equals, it means that the list is sorted.

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/SortNotesTest.java (hunk 9 above)
Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/SortNotesTest.java]: changed it :D

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/SortNotesTest.java (hunk 8 above)
Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/SortNotesTest.java:44]: added the sort by Creation date before adding the random notes

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/NewTextNote.java (hunk 5 above)
Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/NewTextNote.java:49]: i needed to use the clickBack method inside this method, that is why this assertion is here

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/NewCheckList.java (hunk 4 above)
Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/NewCheckList.java]: done

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/EspressoUtils.java (hunk 3 above)
Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/EspressoUtils.java]: done

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/LockedNoteTest.java (hunk 2 above)
Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/LockedNoteTest.java]: done

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/AddSingleTextNoteTest.java (hunk 1 above)
Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/AddSingleTextNoteTest.java:60]: they were moved because i needed the clickback method in the addNewTextNote method

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+        } while (!isSelected());
+        clickOnButton(onView(withText(buttonText)));
+    }
+
+    /**
+     * This method returns true if the button is checked
+     */
+    private static boolean isSelected() {
+        return checkIfMatches(onView(allOf(withId(R.id.radio), withParent(allOf(withId(R.id.content), withParent(childAtPosition(isAssignableFrom(ListView.class), 1)))))), isChecked());
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java]: you use one particular index in this check (index 1), but you call this method above in sortBy method, which is called for multiple buttons (title and creation date), which are at different indexes

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/EspressoUtils.java [hunk 13]
```diff
@@ -90,54 +120,73 @@ static void writeSelectedUiObjectWithText(String textToBeFound, String textToBeR
...
-    }
-
+    /**
+     * this method checks if a text is completely displayed in a view
+     *
+     * @param p is the id used for searching the text
+     * @param s is the text searched
+     * @return true or false - depends if the text is or is not completely displayed
+     */
     public static boolean checkIfTextIsDisplayed(int p, String s) {
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/EspressoUtils.java]: give those parameters a more suggestive name

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/EspressoUtils.java [hunk 14]
```diff
@@ -90,54 +120,73 @@ static void writeSelectedUiObjectWithText(String textToBeFound, String textToBeR
...
             Log.e("add password", "Exception when trying to add password");
         }
     }
 
+    /**
+     * this ViewAction waits a number of milliseconds for a View with a selected content description to be displayed
+     *
+     * @param contentDescription is the param used for the selected content description
+     * @param millis             is the number of milliseconds
+     */
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/EspressoUtils.java:191]: is this taken from somewhere?

Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/EspressoUtils.java:191]: yes, I added the link from where I found it :D

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/EspressoUtils.java (hunk 13 above)
Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/EspressoUtils.java]: done

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java (hunk 12 above)
Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java]: yes, you are right, I added a hashmap which contains the titles of the sorting options...and based on the selected option, will get the index and than check if this particular option is or isn't selected

**Comment by rbutnariu** [comment]
👍 looks good

File: omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+        mapWithSortingOptions.put("Reminder date",3);
+
+
+        ViewInteraction sortButton = onView(withId(R.id.menu_sort));
+        clickOnButton(sortButton);
+        do {
+            clickOnButton(onView(withText(buttonText)));
+            clickOnButton(sortButton);
+
+        } while (!isSelected(mapWithSortingOptions.get(buttonText)));
```
Comment (rbutnariu) [inline omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java:77]: one more thing: this loop can (and will sometimes) loop forever. you need a stop condition. something like a max number of tries. between tries, try to wait 1 second or so to allow the app to process your request.

Comment (rbutnariu) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java:77]: one workaround to avoid repetitive checks and selections is to put the focus on other element after the first button selection before doing again the check, for example: after you selected once creation date or title button, click on some other button (the three-dots button near the sort button) and then redo the check

Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java:77]: > one more thing: this loop can (and will sometimes) loop forever. you need a stop condition. something like a max number of tries. between tries, try to wait 1 second or so to allow the app to process your request.

added a 10 max number of tries and a 2 sec wait between them

Comment (finarugeanina) [reply omniNotes/src/androidTest/java/it/feio/android/omninotes/utils/SortNotes.java:77]: > one workaround to avoid repetitive checks and selections is to put the focus on other element after the first button selection before doing again the check, for example: after you selected once creation date or title button, click on some other button (the three-dots button near the sort button) and then redo the check

added a clicking on the add new notes button (twice - once for open and once for closing)

**Comment by rbutnariu** [comment]
👍
